src/Orleans/CodeGeneration/GrainInterfaceData.cs

```diff
@@ -145,22 +145,102 @@ public static bool IsGrainReference(Type t)
         {
             return typeof(IAddressable).IsAssignableFrom(t);
         }
-        
-        public static MethodInfo[] GetMethods(Type grainType, bool bAllMethods = false)
+
+        //This is to fix issue #504 (which is in turn required to fix issue #492). 
+        //The reason that the previous implementation did not work is the way .Net handles interface "inheritance". 
+        //See: http://stackoverflow.com/questions/10550970/how-to-do-proper-reflection-of-base-interface-methods (last visited: 6/12/15)
+        //See: http://www.roxolan.com/2009/04/interface-inheritance-in-c.html (last visited: 6/12/15)
+        //The type.GetMethods(flags) (where type is an interface, not a class) only returns methods declared by the immediate interface and not by it's parents. 
+        //Hence a recursive solution is used in the following method. 
+        /// <summary>
+        /// Returns all the methods declared by the grain interface and it's parents in the inheritance tree.
+        /// </summary>
+        /// <param name="grainInterfaceType">"grainInterfaceType" must be an interface (not a class) and must be inherited from IGrain at some level.</param>
+        /// <param name="bIncludeAllParentsMethods"></param>
+        /// <returns></returns>
+        public static MethodInfo[] GetGrainInterfaceMethods(Type grainInterfaceType, bool bIncludeAllParentsMethods = false)
+        {
+            if (!IsGrainInterface(grainInterfaceType))
+            {
+                //should never happen since this method should be called only for grain interfaces.
+                throw new ArgumentException("Illegal argument: grainInterfaceType. Must be an interface and extend IGrain");
+            }
+
+            List<MethodInfo> methodInfos = new List<MethodInfo>();
+            GetInterfaceMethods(grainInterfaceType, ref methodInfos, bIncludeAllParentsMethods);
+
+            return methodInfos.ToArray();
+        }
+
+        /// <summary>
+        /// Recurses through entire interface graph accumulating methods.
+        /// </summary>
+        /// <param name="interfaceType"></param>
+        /// <param name="methodInfos"></param>
+        /// <param name="bIncludeAllParentsMethods"></param>
+        private static void GetInterfaceMethods(System.Type interfaceType, ref List<MethodInfo> methodInfos, bool bIncludeAllParentsMethods)
         {
-            var methodInfos = new List<MethodInfo>();
-            GetMethodsImpl(grainType, grainType, methodInfos);
-            var flags = BindingFlags.Public | BindingFlags.InvokeMethod | BindingFlags.Instance;
-            if (!bAllMethods)
-                flags |= BindingFlags.DeclaredOnly;
+            MethodInfo[] infos = interfaceType.GetMethods();
```

Comment:
Given that we are calling this function must be used only for interfaces and not classes, the default flags works fine. (default is BindingFlags.Public). We dont need InvokeMethod flag because interface cannot contain constructors (hence all methods are invokable) and we dont need BindingFlags.Instance flag because interfaces cannot have static methods either.

I will add a explict check that the given type is an interface and also a not on why defaults work.